### PR TITLE
Add support for managing components

### PR DESCRIPTION
**Pull request by git-harry**
This change adds the tool `component` for managing the component
information stored in the releases repository as well as that located in
the component repositories.
    
The tools includes support for the following:
    - registering a component
    - updating a component entry
    - adding a new release
    - tracking component dependencies
    - generating pinned component requirements from dependency constraints
    - downloading requirements

### CODE REVIEW

File: scripts/components.py
Comment (mattt416) [inline scripts/components.py]: Should you ensure the component exists before you try to update it?

File: scripts/components.py
Comment (mattt416) [inline scripts/components.py]: It looks like `--repo-url` is required when subparser is `component`.

**Comment by mattt416** [comment]
@git-harry Haven't gone through the entire PR yet, will continue on later tonight or first thing tomorrow.

File: scripts/dependencies.py
Comment (mattt416) [inline scripts/dependencies.py]: As mentioned in IRC, we probably need a way to create a skeleton component_metadata.yml file without any constraints (here name is required).

File: scripts/dependencies.py
Comment (git-harry) [inline scripts/dependencies.py]: `releases_repo` should be `component["repo_url"]`.

**Comment by mattt416** [comment]
The ordering isn't quite right here, but noting for future reference.

----

#### Register components

```
cd rpc-metadata
python scripts/components.py --component-name rpc-component-1 --component-dir . component --repo-url https://github.com/rcbops/rpc-component-1
python scripts/components.py --component-name rpc-component-2 --component-dir . component --repo-url https://github.com/rcbops/rpc-component-2
python scripts/components.py --component-name rpc-product-1 --component-dir . component --repo-url https://github.com/rcbops/rpc-product-1 --is-product
```

#### Add dependencies to rpc-product-1

```
cd rpc-product-1
python ../rpc-metadata/scripts/dependencies.py --dependency-dir . --releases-repo ../rpc-metadata set-dependency --name rpc-component-1 --constraint "version<r2.0.0"
python ../rpc-metadata/scripts/dependencies.py --dependency-dir . --releases-repo ../rpc-metadata set-dependency --name rpc-component-2 --constraint "version<r2.0.0"
```

#### Create empty component_metadata.yml files in rpc-component-1 and rpc-component-2

The tooling currently does not support creating component metadata for components with no constraints.

```
cd rpc-component-1
echo "dependencies: []" > component_metadata.yml
cd ../rpc-compoment-2
echo "dependencies: []" > component_metadata.yml
```

#### Specify dependency requirements

```
cd rpc-product-1
python ../rpc-metadata/scripts/dependencies.py --releases-repo ../rpc-metadata --dependency-dir . update-requirements --component-dir ../rpc-metadata
```

#### Create component release

```
cd rpc-metadata
python ./scripts/components.py --component-name rpc-component-1 release --version r1.0.0 --sha 540464d96be5cbbdf1cf07ae5457dd57b159c110 --series-name master
python ./scripts/components.py --component-name rpc-component-2 release --version r1.0.0 --sha ed18fd1f7c8bc0db3148ea39a6500cc8f8abb2b6 --series-name master
```

File: scripts/dependencies.py
Comment (mattt416) [inline scripts/dependencies.py]: Indentation of this function is off.

File: scripts/dependencies.py
Comment (mattt416) [inline scripts/dependencies.py]: How should constraints be specified?  If I have a component with two versions (r1.0.0, r1.1.0) and try to update requirements on a component that uses that dependent component, it seems to select the first version of the dependent component that it finds rather than finding the newest version that satisfies the constraint.  I've specified the constraint as `version<r2.0.0` so I'd expect it to select r1.1.0 but it's choosing r1.0.0.

File: scripts/lib/schemata.py
Comment (mattt416) [inline scripts/lib/schemata.py]: We may want to relax this schema to `Regex(r"^https://github.com/.+/.+")` until this code merges into rcbops.

File: scripts/dependencies.py
Comment (mattt416) [inline scripts/dependencies.py]: `repo_dir = dl_base_base_dir + '/' + requirement["name"]`, incase someone doesn't add a trailing slash to `dl_base_dir`.

File: scripts/components.py
Comment (git-harry) [reply scripts/components.py]: Not necessarily but that didn't fail with a nice message so I've added something for now although I may want to rework that area.
```
$ component --releases-dir . release --component-name foobar --version 1.1.1 --series-name blah --sha 0000000000000000000000000000000000000000
Component 'foobar' does not exist.

```

File: scripts/components.py
Comment (git-harry) [reply scripts/components.py]: It wasn't always required, I've split the functionality up to make that clear.

File: scripts/dependencies.py
Comment (git-harry) [reply scripts/dependencies.py]: done

File: scripts/dependencies.py
Comment (git-harry) [reply scripts/dependencies.py]: done

File: scripts/dependencies.py
Comment (git-harry) [reply scripts/dependencies.py]: That should work now. I've added some tests to verify. The list of releases needed reversing. I've also improved the constraints code so something like `version<r2` is now valid.

File: scripts/lib/schemata.py
Comment (git-harry) [reply scripts/lib/schemata.py]: done

File: scripts/dependencies.py
Comment (git-harry) [reply scripts/dependencies.py]: Fixed with `os.path.join`.

**Comment by mattt416** [comment]
Updated notes to reflect latest changes:

----


**Register components**

```
cd rpc-metadata
python rpc-component/component.py --releases-dir=. component --component-name rpc-product-1 add --repo-url https://github.com/rcbops/rpc-product-1
python rpc-component/component.py --releases-dir=. component --component-name rpc-component-1 add --repo-url https://github.com/rcbops/rpc-component-1
python rpc-component/component.py --releases-dir=. component --component-name rpc-component-2 add --repo-url https://github.com/rcbops/rpc-component-2
```

**Add dependencies to rpc-product-1**

```
cd rpc-product-1
python ../rpc-metadata/rpc-component/component.py dependency set-dependency --name rpc-component-1 --constraint "version<r2.0.0"
python ../rpc-metadata/rpc-component/component.py dependency set-dependency --name rpc-component-2 --constraint "version<r2.0.0"
```

**Create dependent component releases**

```
cd rpc-metadata
python ../rpc-metadata/rpc-component/component.py --releases-dir=. release --component-name rpc-component-1 --version 1.0.0 --sha 5773026bea0bd72f8920b7e4f0e13e88b368cd55 --series-name master
python ../rpc-metadata/rpc-component/component.py --releases-dir=. release --component-name rpc-component-2 --version 1.0.0 --sha 540464d96be5cbbdf1cf07ae5457dd57b159c110 --series-name master
```

**Specify dependency requirements**

```
cd rpc-product-1
python ../rpc-metadata/rpc-component/component.py --releases-dir=../rpc-metadata  dependency update-requirements
```

**Create remaining component release**

```
python ../rpc-metadata/rpc-component/component.py --releases-dir=. release --component-name rpc-product-1 --version 1.0.0 --sha e1d1b2348d9eaae3e5a2217372f97a5144b63355--series-name master
```

File: scripts/dependencies.py
Comment (mattt416) [reply scripts/dependencies.py]: Looks like this is still required.  Do we assume if a component has no dependencies then there's no in-repo metadata/requirements files?

**Review by mattt416 — APPROVED** [review]
Need to figure out about the bare dependencies still, but generally speaking this all works so happy to merge and we get more eyes on it when it merges into an rcbops repo.

File: rpc-component/lib/schemata.py
Comment (mattt416) [inline rpc-component/lib/schemata.py]: This not used anywhere?

Comment (git-harry) [reply rpc-component/lib/schemata.py]: There were a couple of unused imports, I've cleaned that up now.

**Pull request merged**
